dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/preProcess/update/ProgramInstanceUpdatePreProcessor.java

```diff
@@ -0,0 +1,150 @@
+
+        Date dueDate = new Date();
+        if ( event.getDueDate() != null )
+        {
+            dueDate = parseDate( event.getDueDate() );
+        }
+
+        if ( event.getEventDate() != null )
+        {
+            programStageInstance.setExecutionDate( parseDate( event.getEventDate() ) );
+        }
+
+        if ( categoryOptionCombo != null )
+        {
+            programStageInstance.setAttributeOptionCombo( categoryOptionCombo );
+        }
+
+        final String storedBy = getValidUsername( event.getStoredBy(), null,
+            importOptions.getUser() != null ? importOptions.getUser().getUsername() : "[Unknown]" );
+
+        final String completedBy = getValidUsername( event.getCompletedBy(), null,
+            importOptions.getUser() != null ? importOptions.getUser().getUsername() : "[Unknown]" );
+
+        if ( event.getStatus() == ACTIVE )
+        {
+            programStageInstance.setStatus( ACTIVE );
+            programStageInstance.setCompletedBy( null );
+            programStageInstance.setCompletedDate( null );
+        }
+        else if ( programStageInstance.getStatus() != event.getStatus() && event.getStatus() == COMPLETED )
+        {
+            programStageInstance.setCompletedBy( completedBy );
+
+            Date completedDate = new Date();
+
+            if ( event.getCompletedDate() != null )
+            {
+                completedDate = DateUtils.parseDate( event.getCompletedDate() );
+            }
+
+            programStageInstance.setCompletedDate( completedDate );
+            programStageInstance.setStatus( COMPLETED );
+        }
+        else if ( event.getStatus() == SKIPPED )
+        {
+            programStageInstance.setStatus( SKIPPED );
+        }
+        else if ( event.getStatus() == SCHEDULE )
+        {
+            programStageInstance.setStatus( SCHEDULE );
+        }
+
+        programStageInstance.setStoredBy( storedBy );
+        programStageInstance.setDueDate( dueDate );
+        programStageInstance.setOrganisationUnit( organisationUnit );
+        programStageInstance.setGeometry( event.getGeometry() );
+
+        if ( programStageInstance.getProgramStage().isEnableUserAssignment() )
+        {
+            programStageInstance.setAssignedUser( ctx.getAssignedUserMap().get( event.getUid() ) );
+        }
+    }
+
+    private String getValidUsername( final String userName, final ImportSummary importSummary,
+        final String fallbackUsername )
+    {
+        String validUsername = userName;
+
+        if ( isEmpty( validUsername ) )
+        {
+            validUsername = getSafeUsername( fallbackUsername );
+        }
+        else if ( validUsername.length() > USERNAME_MAX_LENGTH )
+        {
+            if ( importSummary != null )
+            {
+                // TODO: luciano this should be moved to the new logic
+                importSummary.getConflicts().add( new ImportConflict( "Username", validUsername + " is more than "
```

Comment:
Yes exactly, I have created a shared function for this and a new Validator for the username lenght (note that this is a warning, rather than a breaking error)